cmake/kokkos_arch.cmake

```diff
@@ -574,15 +574,37 @@ CHECK_AMDGPU_ARCH(VEGA906 gfx906) # Radeon Instinct MI50 and MI60
 CHECK_AMDGPU_ARCH(VEGA908 gfx908) # Radeon Instinct MI100
 CHECK_AMDGPU_ARCH(VEGA90A gfx90a) # Radeon Instinct MI200
 
+FUNCTION(SET_AND_CHECK_AMD_ARCH ARCH FLAG)
+  SET(AMD_ARCHITECTURE ${ARCH} PARENT_SCOPE)
+  KOKKOS_SET_OPTION(ARCH_${ARCH} ON)
+  CHECK_AMDGPU_ARCH(${ARCH} ${FLAG})
+ENDFUNCTION()
+
 IF(KOKKOS_ENABLE_HIP AND NOT AMDGPU_ARCH_ALREADY_SPECIFIED)
   IF(KOKKOS_CXX_COMPILER_ID STREQUAL HIPCC)
     FIND_PROGRAM(ROCM_ENUMERATOR rocm_agent_enumerator)
     EXECUTE_PROCESS(COMMAND ${ROCM_ENUMERATOR} OUTPUT_VARIABLE GPU_ARCHS)
     STRING(LENGTH "${GPU_ARCHS}" len_str)
     # enumerator always output gfx000 as the first line
     IF(${len_str} LESS 8)
-      MESSAGE(SEND_ERROR "HIP enabled but no AMD GPU architecture currently enabled. "
-                         "Please enable one AMD GPU architecture via -DKokkos_ARCH_{..}=ON'.")
+      MESSAGE(SEND_ERROR "HIP enabled but no AMD GPU architecture could be automatically detected. "
+                         "Please manually enable one AMD GPU architecture via -DKokkos_ARCH_{..}=ON'.")
+    # check for known gpu archs, otherwise error out
+    ELSE()
+      STRING(REGEX MATCH "(gfx900|gfx906|gfx908|gfx90a)" DETECTED_GPU_ARCH ${GPU_ARCHS})
```

Comment:
What if there are multiple (different) GPUs? I guess it's fine to take the first one since we are also taking just whatever for Cuda in that case.

Reply:
Good question. It basically uses the first valid one returned by `rocm_agent_enumerator`. Unfortunately documentation is sparse and it does not provide a `--help`. Maybe @arghdos can provide some insight on how the output is ordered (e.g. is the gpu with the highest capability listed as first one)

Reply:
But if we know which order we want I can change the regex to enforce a specific order.

Reply:
I'm fine with the current version.

Reply:
> how the output is ordered (e.g. is the gpu with the highest capability listed as first one)

I believe in the past I was told it's enumerating things by PCIe ID.  I.e., this should *probably* match the order it shows up in lspci.  The reason I believe this is because it doesn't have any dependence on any of the runtimes.  It's _possible_ this matched the view of ROCR_VISIBLE_DEVICES (the HSA layer), but I'd have to test this.

Reply:
Ah, ok -- reading the script (it's just a python script) and it documents the ordering: https://github.com/RadeonOpenCompute/rocminfo/blob/master/rocm_agent_enumerator#L225

Basically it chooses the first of:

1. ROCM_TARGET_LIST -- used in docker
2. tgt.list -- used in docker
3. HSA [this is actually using the kernel driver ordering under the hood]
4. lspci
5. rocminfo
To enforce an order, we _could_ create a temporary file matching the regex order, and export ROCM_TARGET_LIST to that file, and remove it after the fact.